File: python/food-chain/food_chain.py
# Define the list of animals in the song
animals = ["fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse"]

# Define phrases associated with each verse
phrases = [
    "It wriggled and jiggled and tickled inside her.",
    "How absurd to swallow a bird!",
    "Imagine that, to swallow a cat!",
    "What a hog, to swallow a dog!",
    "Just opened her throat and swallowed a goat!",
    "I don't know how she swallowed a cow!",
    "She's dead, of course!",
]

# Function to generate the specified range of verses
def recite(start_verse, end_verse):
    ret = []  # Initialize an empty list to store the verses

    # Loop through the requested verses
    for n in range(start_verse, end_verse + 1):
        # Extend the list with the lines of the current verse and an empty line
        ret.extend(verse(n) + [""])

    return ret[:-1]  # Remove the trailing empty line

# Function to generate a single verse
def verse(n):
    ret = ["I know an old lady who swallowed a {}.".format(animals[n - 1])]  # Initialize with the first line of the verse

    # Add additional lines based on the verse number
    if n > 1:
        ret.append("{}".format(phrases[n - 2]))  # Add the associated phrase for the verse
    if n < 8:
        for i in range(n - 1):
            ret.append("She swallowed the {} to catch the {}.".format(
                animals[n - i - 1], animals[n - i - 2]))
            if ret[-1].endswith("spider."):
                ret[-1] = ret[-1][:-1] + " that" + phrases[0][2:]  # Replace "spider." with the spider-related phrase
        ret.append("I don't know why she swallowed the fly. Perhaps she'll die.")  # Add the final line for the verse

    return ret  # Return the lines of the verse as a list

File: javascript/food-chain/food-chain.js
// Define an array representing the animals in the song's food chain
const ANIMALS = [
  'fly', 'spider', 'bird', 'cat',
  'dog', 'goat', 'cow', 'horse',
];

// Define rhymes for specific animals in the food chain
const RHYME = {
  spider: "It wriggled and jiggled and tickled inside her.\n",
  bird: "How absurd to swallow a bird!\n",
  cat: "Imagine that, to swallow a cat!\n",
  dog: "What a hog, to swallow a dog!\n",
  goat: "Just opened her throat and swallowed a goat!\n",
  cow: "I don't know how she swallowed a cow!\n",
  horse: "She's dead, of course!",
}

// Define a function to determine the reason for swallowing an animal
const why = (idx) => {
  if (idx == 0) return "I don't know why she swallowed the fly. Perhaps she'll die.";
  let [That, This] = [ANIMALS[idx - 1], ANIMALS[idx]];
  if (That == 'spider') That += " that wriggled and jiggled and tickled inside her";
  return `She swallowed the ${This} to catch the ${That}.`;
}

// Define a function to create the verses leading up to the animal in the food chain
const ladder = (animal) => {
  if (animal == 'horse') return '';
  let ret = [];
  let idx = ANIMALS.indexOf(animal);
  for (let i = idx; i >= 0; i--) ret.push(why(i));
  return ret.join("\n");
}

// Export a class 'Song' with methods to generate verses and groups of verses
export class Song {
  verse(num) {
    let animal = ANIMALS[num - 1];
    let ret = "I know an old lady who swallowed a ";
    ret += animal + ".\n";
    ret += RHYME[animal] || "";
    ret += ladder(animal) + "\n";
    return ret;
  }

  verses(first, last) {
    let ret = [];
    for (let i = first; i <= last; i++)
      ret.push(this.verse(i));
    return ret.join("\n") + "\n";
  }
}
